### PR TITLE
feat: permissions support

### PR DESCRIPTION
**Pull request by tjrivera**
To be reviewed post #4 

This PR adds support for operations on
* Endpoint permissions
* Entity permissions
* Role operations corresponding to these identities.

### CODE REVIEW

**Comment by hbagdi** [comment]
Please take care of lint failures.

File: kong/kong.go [hunk 1]
```diff
@@ -29,22 +29,24 @@ var (
...
+	Certificates        *CertificateService
+	Plugins             *PluginService
+	SNIs                *SNIService
+	Upstreams           *UpstreamService
+	Targets             *TargetService
+	Workspaces          *WorkspaceService
+	Admins              *AdminService
+	RBACUsers           *RBACUserService
+	EndpointPermissions *EndpointPermissionService
+	EntityPermissions   *EntityPermissionService
```
Comment (hbagdi) [inline kong/kong.go:48]: Let's prepend `RBAC` to both of these new services.

File: kong/types.go [hunk 2]
```diff
@@ -267,3 +270,84 @@ type Role struct {
...
 	IsDefault *bool   `json:"is_default,omitempty" yaml:"is_default,omitempty"`
 }
+
+// PermissionRoleID represents a Role identifier within a permission struct
+type PermissionRoleID struct {
+	ID *string `json:"id,omitempty" yaml:"id,omitempty"`
+}
+
+// EndpointPermission represents an RBAC Endpoint Permisson in Kong Enterprise
+type EndpointPermission struct {
```
Comment (hbagdi) [inline kong/types.go:280]: Please rename this to `RBACEndpointPermission`.

File: kong/types.go [hunk 3]
```diff
@@ -267,3 +270,84 @@ type Role struct {
...
+	CreatedAt *int              `json:"created_at,omitempty" yaml:"created_at,omitempty"`
+	Workspace *string           `json:"workspace,omitempty" yaml:"workspace,omitempty"`
+	Endpoint  *string           `json:"endpoint,omitempty" yaml:"endpoint,omitempty"`
+	Actions   []*string         `json:"actions,omitempty" yaml:"actions,omitempty"`
+	Negative  *bool             `json:"negative,omitempty" yaml:"negative,omitempty"`
+	Role      *PermissionRoleID `json:"role,omitempty" yaml:"role,omitempty"`
```
Comment (hbagdi) [inline kong/types.go:286]: Question: Why can you reference the `RBACRole` struct here? That is how we do foreign-key referencing with other related entities.

File: kong/types.go [hunk 4]
```diff
@@ -267,3 +270,84 @@ type Role struct {
...
+		CreatedAt: e.CreatedAt,
+		Workspace: e.Workspace,
+		Endpoint:  e.Endpoint,
+		Actions:   String(strings.Join(actions, ",")),
+		Comment:   e.Comment,
+	})
+}
+
+// EntityPermission represents an RBAC Entity Permission in Kong Enterprise
+type EntityPermission struct {
```
Comment (hbagdi) [inline kong/types.go:315]: Please rename this to `RBACEntityPermission`.

File: kong/types.go [hunk 5]
```diff
@@ -267,3 +270,84 @@ type Role struct {
...
+	CreatedAt  *int              `json:"created_at,omitempty" yaml:"created_at,omitempty"`
+	EntityID   *string           `json:"entity_id,omitempty" yaml:"entity_id,omitempty"`
+	EntityType *string           `json:"entity_type,omitempty" yaml:"entity_type,omitempty"`
+	Actions    []*string         `json:"actions,omitempty" yaml:"actions,omitempty"`
+	Negative   *bool             `json:"negative,omitempty" yaml:"negative,omitempty"`
+	Role       *PermissionRoleID `json:"role,omitempty" yaml:"role,omitempty"`
```
Comment (hbagdi) [inline kong/types.go:321]: Same question here, why can't this use `RBACRole`?

File: kong/types.go [hunk 6]
```diff
@@ -267,3 +270,84 @@ type Role struct {
...
+		CreatedAt:  e.CreatedAt,
+		EntityID:   e.EntityID,
+		EntityType: e.EntityType,
+		Actions:    String(strings.Join(actions, ",")),
+		Comment:    e.Comment,
+	})
+}
+
+// PermissionsList is a list of permissions, both endpoint and entity, associated with a Role.
+type PermissionsList struct {
```
Comment (hbagdi) [inline kong/types.go:350]: Please rename this to `RBACPermissionsList`.

File: kong/types.go [hunk 7]
```diff
@@ -267,3 +270,84 @@ type Role struct {
...
+		EntityType *string           `json:"entity_type,omitempty" yaml:"entity_type,omitempty"`
+		Actions    *string           `json:"actions,omitempty" yaml:"actions,omitempty"`
+		Negative   *bool             `json:"negative,omitempty" yaml:"negative,omitempty"`
+		Role       *PermissionRoleID `json:"role,omitempty" yaml:"role,omitempty"`
+		Comment    *string           `json:"comment,omitempty" yaml:"comment,omitempty"`
+	}
+	var actions []string
+	for _, action := range e.Actions {
+		actions = append(actions, *action)
+	}
```
Comment (hbagdi) [inline kong/types.go:339]: Question: Why is this required? Why can't we mirror what is present in Kong's Admin API?

**Review by hbagdi — CHANGES_REQUESTED** [review]

File: kong/types.go (hunk 7 above)
Comment (tjrivera) [reply kong/types.go:339]: Unfortunately the way that Kong's Admin API accepts POSTs and PUTs is in a different form (comma-separated vs. actual array) than what you receive from a GET request.

File: kong/types.go (hunk 5 above)
Comment (tjrivera) [reply kong/types.go]: I suppose we can -- marshaling would just omit the empty fields.

File: kong/endpoint_permission_service.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+)
+
+// RBACEndpointPermissionservice handles RBACEndpointPermissions in Kong.
+type RBACEndpointPermissionservice service
```
Comment (hbagdi) [inline kong/endpoint_permission_service.go:11]: `RBACEndpointPermissionservice` should be `RBACEndpointPermissionService`.

File: kong/entity_permission_service.go [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+)
+
+// RBACEntityPermissionservice handles RBACEntityPermissions in Kong.
+type RBACEntityPermissionservice service
```
Comment (hbagdi) [inline kong/entity_permission_service.go:11]: `RBACEntityPermissionservice` should be `RBACEntityPermissionService`

File: kong/endpoint_permission_service.go [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+// Create creates a RBACEndpointPermission in Kong.
+func (s *RBACEndpointPermissionservice) Create(ctx context.Context,
+	ep *RBACEndpointPermission) (*RBACEndpointPermission, error) {
+
+	if ep == nil {
+		return nil, errors.New("cannot create a nil endpointpermission")
+	}
+	if ep.Role.ID == nil {
```
Comment (hbagdi) [inline kong/endpoint_permission_service.go:20]: You can have a nil pointer panic here.
Please do `ep.Role == nil || ep.Role.ID == nil`.
Please change this in all places.

**Review by hbagdi — APPROVED** [review]
Did a rough pass, let's iterate if there are bugs in here.
